<a href="https://colab.research.google.com/github/ttogle918/ds-section1-sprint1/blob/master/n112a_feature_engineering_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src='https://user-images.githubusercontent.com/6457691/89746244-a6e2a200-daf3-11ea-9314-7863d544d73c.png' width = '200' align = 'right'>

## *DATA SCIENCE / SECTION 1 / SPRINT 1 / NOTE 2*

# 📝 Assignment


## 1. Feature Engineering

- `KT&G` 기업 데이터를 계속 사용합니다.
  - 변수 이름은 `df`로 저장하세요.

- 각 분기별 영업이익률을 직접 계산해보세요. 
  - Feature 이름은 `영업이익률2` 입니다.

- 아래 데이터를 참조하여 18/19년도에 대해서 Price to Dream Ratio (PDR)을 계산하세요.
  - 변수 이름은 `df2`, feature 이름은 `PDR`로, `df2['PDR']`을 통해 `~` 값이 확인 되어야합니다. 아래의 전제조건을 활용하세요.
  - 점유율 (일반담배 기준) 
    - 18년 : 62%
    - 19년 : 63.5% 
  - 시장의 규모 
     - 18년 : 15조 
     - 19년 : 15조

||구분|종가|발행주식수|시가총액|시장점유율|PDR|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|0|18년|101500|137292497|13935188445500|62.0|~|
|1|19년|93800|137292497|12878036218600|63.5|~|


이상치
잘못된 값이 아니라면.. 모두 다 평균범주는 아닐 수 있다.. 실제 값의 범위는 넓다.?
잘못입력된 값 혹은 평균 범주를 벗어난 값



In [19]:
import pandas as pd
from pandas import Series, DataFrame

url = 'https://ds-lecture-data.s3.ap-northeast-2.amazonaws.com/kt%26g/kt%26g_0.csv'
headers = ['분기', '매출액', '영업이익', '영업이익(발표기준)', '세전계속사업이익',
           '당기순이익', '당기순이익(지배)', '당기순이익(비지배)', '자산총계', '부채총계',
           '자본총계', '자본총계(지배)', '자본총계(비지배)', '자본금', '영업활동현금흐름',
           '투자활동현금흐름', '재무활동현금흐름', '영업이익률', '순이익률', 'ROE(%)',
           'ROA(%)', '부채비율', '자본유보율', 'EPS(원)', 'PER(배)']
def toInt(str1):
  if type(str1) == int or type(str1) == float:
    return str1
  return int(str1.replace(',', ''))

def toFloat(str1):
  if type(str1) == float :
    return str1
  return float(str1.replace(',', ''))
df = pd.read_csv(url, names=headers)

# 영업이익률 = 영업이익/매출액 * 100
df['영업이익률2'] = df['영업이익'].apply(toInt) / df['매출액'].apply(toInt) * 100

# TAM : 전체 시장 규모
# PDR = 기업가치(시가총액) / (TAM * 시장점유율)
# 기업가치(EV) = 시가총액(MC) + 총부채(Total Debt) - 현금 및 현금성 자산(C)
# 시가총액 = 자본총계
df2 = DataFrame({'구분': ['18년', '19년'], '종가': [101500, 93800], '발행주식수': [137292497, 137292497], 
                 '시가총액': ['13935188445500', '12878036218600'], '시장점유율':[62.0, 63.5], '시장의규모' :[15000000000000, 15000000000000]})


df2['PDR'] = df2['시가총액'].apply(toInt) / ( df2['시장의규모'] + df2['시장점유율']) 
# data를 정수(숫자)로 입력했는데, Int, float로.. type캐스팅 할려고 하니까 계속 에러가 났음.
df

분기     매출액   영업이익 영업이익(발표기준)  ...   자본유보율 EPS(원) PER(배)     영업이익률2
0   19/6  12,578  3,991      3,991  ...  803.22  2,351  13.77  31.730005
1   19/9  13,222  3,825      3,825  ...  836.86  2,337  13.87  28.929058
2  19/12  11,982  2,523      2,523  ...  850.82    873  12.42  21.056585
3   20/3  11,784  3,150      3,150  ...  823.16  2,134   9.71  26.731161
4   20/6  13,188  3,947      3,947  ...  837.40  2,135  10.46  29.928723

[5 rows x 26 columns]

## 2. String Manipulation

- replace가 아닌 다른 방법을 사용하여 19년도 2분기의 매출액 을 integer로 변경하세요. 결과값은 12578이 나와야합니다.
  


In [13]:
# 이 곳에 답안을 작성하시길 바랍니다
def removeComma(string) :
  newstr = ""
  start = 0
  f = string.find(',')
  newstr = string[0:f]
  while f > 0 :
    f = string[f+1:].find(',') 
    newstr += oldstr[:finded]
    start = finded+1
  newstr = newstr + oldstr[finded+1:]
  return newstr

def findYear(string) :
  if int(string[:2]) == 19 :
    return True
  return False

# li = df.loc[df['분기'].apply(findYear) == True, ['매출액']].sum(axis=1) # 분기별
num1 = df.loc[df['분기'] == '19/6', ['매출액']]
num1['매출액'] = num1['매출액'].apply(toInt)

num1
# dataframe -> numeric : 숫자형으로 바꾸기


매출액
0  12578

## 3. Apply
- 데이터셋의 분기를 제외한 모든 feature에 대해서 수치형 데이터 (`int`, `float`) 로 수정하고, `describe`를 통해 summary statistics 를 확인하세요.
- 단 결측치는 `0`으로 처리합니다.

------
df[i].apply(toFloat) 했을 때 dstype으로 확인하면 object 타입이 나온다.

df[i]에 적용이 안된것!

df[i] = df[i].apply(toFloat)로 해야한다.


df.describe() : 평균, 표준 및 IQR값을 제공
문자열은 제외하고 숫자 열에 대해서만 요약 통계를 계산

그래서 describe()를 하기 위해서는 수치형 데이터로 타입 변환을 해야한다!

물론 문자열에 대해서도 요약 통계가 가능하다 -> df.describe(include=['object'])

혹은 모든 열에 대해서 요약 통계도 가능하다 -> df.describe(include='all')

float32로 바꾸려면 어떻게 해야할까?

In [18]:
import numpy as np
# 이 곳에 답안을 작성하시길 바랍니다
df = df.fillna('0')
# print(df.dtypes)
# %(백분율) 만 float형식. 자본총계(비지배)는 float형식이었지만, 원으로 끝난다는 가정하에 int로 변환.
for i in headers :
  if i == '분기' :
    continue
  
  if i == '영업이익률' or i == '순이익률' or i == 'ROE(%)' or i == 'ROA(%)' or i == '부채비율' or i == '자본유보율' or i == 'PER(배)' or i == '영업이익률2' :
    df[i] = df[i].apply(toFloat)
    continue
  
  df[i] = df[i].apply(toInt)

df
# df = df.astype({'매출액':int, '영업이익':int, '영업이익(발표기준)':int, '세전계속사업이익':int,
#            '당기순이익':int, '당기순이익(지배)':int, '당기순이익(비지배)':int, '자산총계':int, '부채총계':int,
#            '자본총계':int, '자본총계(지배)':int, '자본총계(비지배)':int, '자본금':int, '영업활동현금흐름':int,
#            '투자활동현금흐름':int, '재무활동현금흐름':int, '영업이익률':np.float, '순이익률':np.float, 'ROE(%)':np.float,
#            'ROA(%)':np.float, '부채비율':np.float, '자본유보율':np.float, 'EPS(원)':int, 'PER(배)':np.float})

print(df.dtypes)
print('------------------------------------------------------------------------------')
print(df.describe())


분기             object
매출액             int64
영업이익            int64
영업이익(발표기준)      int64
세전계속사업이익        int64
당기순이익           int64
당기순이익(지배)       int64
당기순이익(비지배)      int64
자산총계            int64
부채총계            int64
자본총계            int64
자본총계(지배)        int64
자본총계(비지배)     float64
자본금             int64
영업활동현금흐름        int64
투자활동현금흐름        int64
재무활동현금흐름        int64
영업이익률         float64
순이익률          float64
ROE(%)        float64
ROA(%)        float64
부채비율          float64
자본유보율         float64
EPS(원)          int64
PER(배)        float64
영업이익률2        float64
dtype: object
------------------------------------------------------------------------------
                매출액         영업이익  ...     PER(배)     영업이익률2
count      5.000000     5.000000  ...   5.000000   5.000000
mean   12550.800000  3487.200000  ...  12.046000  27.675106
std      664.979097   636.931079  ...   1.898112   4.116711
min    11784.000000  2523.000000  ...   9.710000  21.056585
25%    11982.000000  3150.000000  .

## 🔥 도전과제



### NA Value Handling

- 19년도 4분기의 당기순이익(비지배) 부분을 Na로 대체하세요
- 이후 해당 결측치를 mean imputation 방법을 사용하여 처리하세요.



In [ ]:
# 이 곳에 답안을 작성하시길 바랍니다

### Feature Engineering

Relative Perfomance 라는 새로운 feature를 계산하세요.

이는 최근 1년치 매출액의 평균값을 기준으로 

- 10% 이상 -> `S`
- 5% 이상 -> `A`
- -5 ~ 5% -> `B`
- -5%이하 -> `C`
- 10%이하 -> `D`

라는 값을 갖는 feature입니다.

20년도 2분기에 해당하는 결과값은 `A`가 나와야합니다.

각각에 해당하는 등급이 나오기 위해서 필요한 매출액을 추가로 서술하세요.

In [ ]:
# 이 곳에 답안을 작성하시길 바랍니다